core/services/ocr2/plugins/ocr2vrf/coordinator/coordinator.go

```diff
@@ -0,0 +1,426 @@
+			}
+			randomnessFulfillmentRequestedLogs = append(randomnessFulfillmentRequestedLogs, unpacked)
+		case bytes.Equal(lg.EventSig, c.randomWordsFulfilledTopic[:]):
+			unpacked, err2 := unmarshalRandomWordsFulfilled(lg) // FIXME
+			if err2 != nil {
+				// should never happen
+				err = errors.Wrap(err2, "unmarshal RandomWordsFulfilled log")
+			}
+			randomWordsFulfilledLogs = append(randomWordsFulfilledLogs, unpacked)
+		case bytes.Equal(lg.EventSig, c.newTransmissionTopic[:]):
+			unpacked, err2 := unmarshalNewTransmission(lg) // FIXME
+			if err2 != nil {
+				// should never happen
+				err = errors.Wrap(err2, "unmarshal NewTransmission log")
+			}
+			newTransmissionLogs = append(newTransmissionLogs, unpacked)
+		}
+	}
+
+	// Scan for blocks where an output is required
+	// blocksRequested maps block number to the block object.
+	type key struct {
+		blockNumber uint64
+		confDelay   uint32
+	}
+	blocksRequested := make(map[key]struct{})
+	for _, r := range randomnessRequestedLogs {
+		if _, ok := confirmationDelays[uint32(r.ConfDelay.Uint64())]; !ok {
+			// if we can't find the conf delay in the map then just ignore this request
+			c.lggr.Infow("ignoring bad request, unsupported conf delay",
+				"confDelay", r.ConfDelay.String(),
+				"supportedConfDelays", confirmationDelays)
+			continue
+		}
+		// If the next beacon output height is less than currentHeight - conf delay
+		// AND the log has enough confirmations, then we can schedule it to be fulfilled.
+		cond := r.ConfDelay.Int64() < currentHeight // TODO: is this necessary? Won't this always be true?
+		cond = cond && r.NextBeaconOutputHeight < uint64(currentHeight-r.ConfDelay.Int64())
+		cond = cond && currentHeight >= int64(r.Raw.BlockNumber+r.ConfDelay.Uint64()) // TODO: is this redundant?
+		if cond {
+			blocksRequested[key{
+				blockNumber: r.NextBeaconOutputHeight,
+				confDelay:   uint32(r.ConfDelay.Uint64()),
+			}] = struct{}{}
+		}
+	}
+
+	// Scan for blocks where a callback is requested
+	var callbacksRequested []vrf_wrapper.VRFBeaconCoordinatorRandomnessFulfillmentRequested
+	for _, r := range randomnessFulfillmentRequestedLogs {
+		if _, ok := confirmationDelays[uint32(r.ConfDelay.Uint64())]; !ok {
+			// if we can't find the conf delay in the map then just ignore this request
+			c.lggr.Infow("ignoring bad request, unsupported conf delay",
+				"confDelay", r.ConfDelay.String(),
+				"supportedConfDelays", confirmationDelays)
+			continue
+		}
+
+		// If the next beacon output height is less than currentHeight - conf delay
+		// AND the log has enough confirmations, then we can schedule it to be fulfilled.
+		cond := r.ConfDelay.Int64() < currentHeight // TODO: is this necessary? Won't this always be true?
+		cond = cond && r.NextBeaconOutputHeight < uint64(currentHeight-r.ConfDelay.Int64())
+		cond = cond && currentHeight >= int64(r.Raw.BlockNumber+r.ConfDelay.Uint64()) // TODO: is this redundant?
```

Comment:
Can be extracted to a private function (e.g. `isBlockReady()`) to avoid duplication in line 207